### ISSUE TITLE
Autogenerate enums for xml files at compile time

### ISSUE THREAD

**Issue by Nightinggale** [We-the-People-civ4col-mod/Mod#58]
I propose using a script I wrote for M:C, which generates enum values to contain the Types from xml.

What it does is it takes a list of strings (xml file names), reads the Types and prints them to enums. This is saved to a temp file and then copied to the header file if the files differs. This way the header file is only touched if something changed which in turn prevents forcing recompilations for no reason at all. Enums are part of the precompiled header file, meaning a change forces all cpp files to be recompiled.

It allows faster performance because the precompiler is aware of values, both for unrolling loops and for stuff like "what is the index for fishing boats?" (the latter took up 1% of the AI CPU time before I cached it). Using enum values supports switch cases, unlike cached values.

The tradeoff from hardcoding is that it will cause bugs if xml files are modified in any way, which affects the enums. Because of this the script inserts ifdef (preprocessor flag), which allows switching between fast hardcoding and relying on xml values, reducing the need for recompiling for people who edit the xml files.

M:C also has autogenerated code containing the md5 checksums of the hardcoded xml files. In hardcoded state, it can detect changes and throw an error to tell people to use the non-hardcoded for modding.

Using an instance of an enum (like YieldTypes) as variables will make the enum string value show up in the debugger instead of the int value. This makes it useful to have enums for debugging even if the compiler doesn't use the enum values.

The script is easy to configure for special rules for files. This is mainly for a few files/enums, which doesn't follow the naming convention. It can also force files to always be hardcoded, like YieldTypes and UnitAI.

The performance of the enum generating script is so good that using the makefile to check dependencies (all the xml files) takes longer than running the script itself. Because of this the script is designed to run each time even if nothing changed.

Semi related: I have code, which allows using ++ as prefix and postfix on instances of enum types. This allows changing:
```cpp
for (int iYield = 0; iYield < NUM_YIELD_TYPES; ++i)
{
      YieldTypes eYield = (YieldTypes)iYield;
```

into

```cpp
for (YieldTypes eYield = FIRST_YIELD; eYield < NUM_YIELD_TYPES; ++eYield)
```

For full support of this, the enum generating script ends the enums with FIRST_(type) = 0. This will help to get rid of typecasting, which is always a good idea for more reasons than just code readability.

The code looks like this (copy pasted from M:C develop branch)

```cpp
// prefix
template <class T>
static inline T& operator++( T& c )
{
	c = (T)(c + 1);
	return c;
}

// postfix
template <class T>
static inline T operator++( T& c, int )
{
	T cache = c;
	c = (T)(c + 1);
	return cache;
}
```

**Comment by ShadesOT** [comment]
I like that like a bee likes honey and flowers.

**Comment by ShadesOT** [comment]
Question1: can eYield be used to index arrays without a cast? (Maybe the cast is implicit.)


Fact 1:
I don't yet fully understand this. But ... well I will see in time. 
> The tradeoff from hardcoding is that it will cause bugs if xml files are modified in any way, which affects the enums. Because of this the script inserts ifdef (preprocessor flag), which allows switching between fast hardcoding and relying on xml values, reducing the need for recompiling for people who edit the xml files.

**Comment by Nightinggale** [comment]
One important fact to add here is that the script assumes [Strawberry perl](http://strawberryperl.com/) to be installed when compiling. In other words it's an additional requirement for compiling.

One nice bonus from adding that requirement is that it also provides GNU make, which is vastly superior to nmake. In fact the makefile could be written a whole lot better if it wasn't for stupid nmake limitations.

**Comment by abmpicoli** [comment]
I personally think that, instead of enums, we should simply try to make a better usage of the enums.  I don't like the idea of having to recompile the dll every time we change an xml. The time to market for a xml file is instant. The time to market for a hardcoded enum is a full rebuild of the solution... Also, the expertise involved with changing a xml and actually fully compile the DLL is totally different...

**Comment by abmpicoli** [comment]
Notice that if we hardcode the xml, the player that want to private mod this values will be unable to do so, unless he also compiles the DLL (which is a pain for a casual player simply wanting to play with the xml files).

**Comment by abmpicoli** [comment]
I believe we should instead refactor the use of xml at the classes so the values needed are fetched only once, either by lazy initialization or by loading during object construction.

**Comment by abmpicoli** [comment]
One example that comes to mind is that one player changed the xml values for food needed for growth to have a better balance (in their opinion) of food x crosses. https://forums.civfanatics.com/threads/food-food-and-more-food.627933/#post-15024220

**Comment by Nightinggale** [comment]
An example of an autogenerated enum in M:C:

```cpp
enum EraTypes
{
	NO_ERA = -1,

#ifdef HARDCODE_XML_VALUES

	ERA_DISCOVERY,
	ERA_EXPANSION,
	ERA_COLONIAL,
	ERA_INDEPENDENCE,

	NUM_ERA_TYPES,
#else //HARDCODE_XML_VALUES
#define NUM_ERA_TYPES (static_cast<EraTypes>(GC.getNumEraInfos()))
#endif // HARDCODE_XML_VALUES
	FIRST_ERA = 0,
};
```

It's fairly simple. If HARDCODE_XML_VALUES is defined at compile time, the enum will be hardcoded. However if HARDCODE_XML_VALUES isn't set, then we are reduced to NO_ERA, FIRST_ERA and NUM_ERA_TYPES. The first two are fixed regardless of the xml file while the latter will actually look up xml data.

This makes the hardcoding optional. In most cases you do want the hardcoded enum while debugging as it helps a lot with both the display and it avoids the calls to get*Infos(), the latter quickly gets annoying.

So what about using hardcoded values when not hardcoding? It's actually not that complex either. Say we need ERA_COLONIAL somewhere. Hardcoded is easy, but for non-hardcoded, we add the following:

```cpp
#ifndef HARDCODE_XML_VALUES
extern EraTypes ERA_COLONIAL;
#endif
```

Then in some cpp file we add the following in a header file:

```cpp
#ifdef HARDCODE_XML_VALUES
void someInitFunction()
{
}
#else
EraTypes ERA_COLONIAL;

void someInitFunction()
{
      ERA_COLONIAL = GC.getDefineINT("ERA_COLONIAL"); // we might consider something to aim it at a specific file to be more clean
}
#endif
```

Now if someInitFunction() is called during xml loading, ERA_COLONIAL can be used like a hardcoded value everywhere. It's easy to check if it's coded right. Non-hardcoded fails if a hardcoded value is used, which isn't turned into a global variable. If a global variable is modified by mistake, the hardcoded version will fail to compile because you try to assign a value to an enum value.

In M:C, Release and Profile DLLs hardcode, Assert doesn't and there are two Debug, with and without hardcoding. If you want to edit xml files, you want to use an Assert build because it checks for all sorts of invalid data (xml or otherwise), which are skipped in Release for performance reasons.

It should be noted that vanilla has some hardcoded enums, most noteworthy YieldTypes, though there are multiple. They will benefit a lot from being autogenerated because you just add to xml and then you compile. Right now you have to update both at the same time manually. The script is aware of which files have to be hardcoded and it will not insert the ifdef in those.

**Comment by abmpicoli** [comment]
Notice that for all that you are assuming that people are super proficient in C++ on average, and aren't intimidated by having to compile the code themselves... I myself felt a little intimidated at the start, and notice that I program since 1985... To read the C++ code for reference is one thing... To actually compile the code was almost 4 hours of preparation and frustrations for me... and 2 days... 

So, if we go hardcoding stuff, it must be only for those things that are really *impossible* a casual modder would want to play with... And I think the scope of those is very little.

I think that things that affect game balance shouldn't be hardcoded, at all: power players may want to change those to experiment them...

**Comment by abmpicoli** [comment]
Oh, I hear what you are saying now... You want the hardcoded enums for C++ debugging purposes only? If that is the case, I have no quarrel with that... As long as the hardcoded values don't get into the final product sent to the community...

**Comment by Nightinggale** [comment]
For releases to the public, we have two options:

1. don't hardcode
2. release two DLL files

1 is simple while 2 would require another script I have written. It makes an md5 checksum of every single hardcoded xml file and compiles this value into the DLL. At startup, it will then calculate the md5 checksums and give an alert if the file is modified. It can then tell some message about using the other DLL or something.

I can't tell right now which option I would like to go for. For the time being, I think we should allow the code to support both options and when we need to make a decision, we should profile the code to tell how big the difference is in performance and make a judgment based on that.

We should also consider a hybrid, like if it's not hardcoded, it still throws an alert if the always hardcoded data has been modified in xml. The enum generating script already has every single Type for each file. We can write C++ header files with them in any way we want.

**Comment by abmpicoli** [comment]
Ok. In that case, I think we can go for it... Changing to :+1:  .

**Comment by abmpicoli** [comment]
@ShadesOT I'm not sure what you mean... if you add new properties to an existing xml schema, you need code to handle it, or course. But you can, for example, add new unit definitions, new buldings... also you *should* be able to add or remove common goods from the game without extra C++ coding.

I also believe it is possible to add extra game speeds... change the default map sizes for each size (tiny , small, medium, large, huge)... 

Not to mention that it is possible to change terrain yields...

**Comment by ShadesOT** [comment]
oh yes, I did not see it. One could add a new common yield and handle it all from the xml: terrain yields, special unit capabilities, buildings, selling ...
I was more stuck in mind with new special yields and new game mechanics.

**Comment by ShadesOT** [comment]
#213 would make this issue obsolete.

**Comment by Nightinggale** [comment]
Implemented in 2.8 branch. XML hardcoding can be turned on or off at compile time. Dll files with hardcoding will check xml files against expectations and display an error if compile time assumptions are no longer valid.
